### ISSUE TITLE
Remove CBC cipher suites from modern config

### ISSUE THREAD

**Issue by jamieburchell** [mozilla/server-side-tls#244]
I recently noticed that a Qualys scan listed two cipher suites as being weak. I was puzzled at first, because the names are slightly different to what this tool outputs, but I believe as per the thread linked below `ECDHE-RSA-AES256-SHA384` (`TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384`) and `ECDHE-RSA-AES256-SHA256` (`TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256`) should be removed:

https://community.qualys.com/message/46538-re-weak-ciphers-listed-in-report-that-are-not-implemented?commentID=46538&et=watches.email.thread#comment-46538

Config used: Apache 2.4.6, Modern, OpenSSL 1.0.2k

**Comment by tomato42** [comment]
browser? yes
library? no: GnuTLS in RHEL-6/CentOS-6 can't do AES-GCM in TLS 1.2

**Comment by tylerknott** [comment]
For reference, [here](https://blog.qualys.com/technology/2019/04/22/zombie-poodle-and-goldendoodle-vulnerabilities) is the blog post where Qualys explains the changes to the SSL Labs grading. The gist is that several new POODLE variants [were presented](https://i.blackhat.com/asia-19/Fri-March-29/bh-asia-Young-Zombie-Poodle-Goldendoodle-and-How-TLSv13-Can-Save-Us-All.pdf) recently at Black Hat Asia 2019 which was enough for them to classify all CBC-mode ciphersuites as "weak".

Also, checking browser compatibility on SSL Labs, Chrome would need to be bumped to 31 (up from 30) and IE 11 on Win 7/8 supports ECDHE+ECDSA or DHE+RSA for AES-GCM but not ECDHE+RSA. This means that if the "modern" config is updated to remove the CBC ciphersuites a note should probably be added to the wiki page emphasizing that IE 11 on Windows 7 and 8 would require an ECDSA certificate to function with the recommended ciphersuites or a non-EC DHE AES-GCM ciphersuite should be added if using an RSA certificate is necessary.

**Comment by jamieburchell** [comment]
Looks like if I read that correctly, they are going to start grading sites that use them with an "F" from the end of the month.

As someone who's not an expert on the subject, it's quite frustrating to know what we are supposed to be doing about it. All I know is that I and presumably many people use this tool to grab the currently recommended modern list of cipher suites to use.

According to Let's Encrypt, "ECDSA Root and Intermediates" isn't being rolled out until Q3, but I'm not sure if that means ECDSA certificates aren't available until then. "ECDSA Signing Support" is apparently already available.

https://letsencrypt.org/upcoming-features/

Is there a reason "non-EC DHE AES-GCM" suites aren't added to the suggested list?

**Comment by tomato42** [comment]
the qualys blog post says

> SSL Labs will start giving “F” grade to the server affected by these vulnerabilities

the key word is _affected_. Using CBC ciphers is not a vulnerability in and out of itself, Zombie POODLE, etc. are the vulnerabilities. Vulnerabilities caused by bugs in implementations.

furthermore
> cipher suites using CBC with orange color [are marked] with text WEAK. This change **won’t have any effect on the grades**, as it only means that SSL Labs discourages the use of CBC-based cipher suites further.

[emphasis mine]

**Comment by tylerknott** [comment]
> According to Let's Encrypt, "ECDSA Root and Intermediates" isn't being rolled out until Q3, but I'm not sure if that means ECDSA certificates aren't available until then. "ECDSA Signing Support" is apparently already available.

Lets Encrypt will sign an ECDSA end-entity certificate using their RSA intermediate certificate. That should be sufficient for IE 11 compatibility. The ECDSA root and intermediate certificate is only needed if your TLS implementation doesn't support RSA at all.

> Is there a reason "non-EC DHE AES-GCM" suites aren't added to the suggested list?

Non-Elliptic Curve DHE (also called "finite field" DHE) is generally slow and in TLS 1.2 and below is vulnerable to weak groups. TLS 1.3 requires specific highly-scrutinized DHE groups (which are 2048-bits and longer) to be used whereas older version of TLS did not specify any restrictions on the groups to use. Many older TLS implementations used very weak DHE groups shorter than 1024-bits, allowed completely invalid groups to be used, or re-used specific 1024-bit groups in a way that could make them a target for factoring attacks (see Logjam). These issues don't impact ECDHE which is why it's generally preferred today.

**Comment by jamieburchell** [comment]
Thanks for explaining that. Well that's quite a conundrum to solve it seems.

My environment is CentOS 7.6 which ships with Apache 2.4.6 and that does not support using a dhparam file. It can be appended to a cert, but I'm using Certbot to obtain and renew certificates automatically. And Certbot doesn't issue ECDSA certificates.

Edit: For anyone else stuck on this, [RH back-ported DHE functionality](http://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslcertificatefile) to Apache 2.4.6. Replacing the CBC suites with the DHE+RSA suites gets us support for IE11 on Win7/8 without having to generate anything else.

```
TLS_DHE_RSA_WITH_AES_256_GCM_SHA384 (0x9f)   DH 2048 bits   FS | 256
TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 (0x9e)   DH 2048 bits   FS | 128
```

In nginx (1.16.0) it's necessary to generate dhparams (at leat 2048 bits) before those suites will work.

**Comment by tomato42** [comment]
> My environment is CentOS 7.6 which ships with Apache 2.4.6 and that does not support using a dhparam file. It can be appended to a cert, but I'm using Certbot to obtain and renew certificates automatically. And Certbot doesn't issue ECDSA certificates.

won't apache 2.4 on CentOS actually scale the DH params used with the RSA key in question? i.e. if the RSA key is 3072 bit long, it will use 3072 bit long parameters?  For RHEL there's a fixed [RHBZ#1180223](https://bugzilla.redhat.com/show_bug.cgi?id=1180223)

**Comment by jamieburchell** [comment]
> won't apache 2.4 on CentOS actually scale the DH params used with the RSA key in question? i.e. if the RSA key is 3072 bit long, it will use 3072 bit long parameters? For RHEL there's a fixed RHBZ#1180223

Yes, I've later discovered that's the case. I always find it confusing when things are back-ported and you check the version you have with the docs. I think though that means it uses pre generated known params and you are encouraged to create your own? Still, at least it works.

According the Qualys guys and your good self, unless the server is vulnerable, I need not take any action anyway.

**Comment by tomato42** [comment]
> I think though that means it uses pre generated known params and you are encouraged to create your own?

you need to generate your own *only* when you absolutely need compatibility with old, unpatched Java clients (like java 6 or 7); in general there is no need (as in, you are *not* encouraged)  to generate your own (if you don't know how to verify the size of used parameters I'd strongly recommend not doing that – it's easy to misconfigure)

**Comment by jamieburchell** [comment]
> you are _not_ encouraged

OK. I think my interpretation of the documentation wasn't as it was intended

http://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslcertificatefile

>  it is recommended to either use custom DH parameters (preferably)

Edit: Other sources I read to back up my interpretation:

>  But they should be "clean", generated by you so you know they're not provided by an attacker.

> This suggests that there is some amount of increased safety to be had in generating your own primes (rather than relying on those that come with your software), and perhaps in re-generating those primes periodically.

https://security.stackexchange.com/a/38207

**Comment by tomato42** [comment]
yes, if you need 1024 bit DH parameters, you really, really should generate your own

> http://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslcertificatefile

AFAICS, it recommends generating your own _when you use openssl 1.0.1 or earlier_, that translates to RHEL-6 or a distribution of similar vintage... for one, few of those will ship apache 2.4

if you can use larger parameters, then the danger of you messing up is much larger than the danger from precomputation attacks against 2048 bit (or higher) primes, instead of using built-in mechanism to automatically select appropriate ones

secondly, using custom parameters will make the server non-compliant with FIPS requirements:

https://csrc.nist.gov/News/2018/NIST-Publishes-Updates-to-SP-800-56A-and-800-56C
> The publication approves the use of specific safe-prime groups of domain parameters for the finite field DH and MQV schemes, in addition to the previously approved domain parameter sets.

in particular, see Table 25 and 26 on page 133 of
https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf

finally, custom DH parameters aren't even a thing in TLS 1.3

**Comment by jamieburchell** [comment]
OK thanks for explaining! I did need to generate them for Nginx, as a connection could not be established when enabling the DHE suites without it. I think it used 1024 bit params by default.

**Comment by tomato42** [comment]
so looks like there is nothing to do for the guide, closing

**Comment by polarathene** [comment]
> GnuTLS in RHEL-6/CentOS-6 can't do AES-GCM in TLS 1.2

[This RHEL6 document](https://access.redhat.com/articles/1493183) (last updated June 2018, so RHEL 6.9, close to RHEL 6.10) states the version of GnuTLS is `2.12.23`. That version is noted in [this Gitlab Nov 2016 milestone](https://gitlab.com/gnutls/gnutls/-/milestones/9) that it was the last version before their `3.0.0.` ABI break. [That resulted in a new point release that was sponsored by Red Hat](https://lists.gnupg.org/pipermail/gnutls-devel/2016-November/008220.html), no clue if that is used in RHEL 6, based on the linked RHEL 6 document, I guess not. Doesn't appear to make any mention of AES-GCM support to the available TLS 1.2 in GnuTLS however.

According to the [GnuTLS changelog](https://fossies.org/linux/gnutls/NEWS) GCM added in `2.99.1` (April 2011), AES-256-GCM supported in `3.0.1` (Aug 2011). `2.12.2` was released in April 2011 as well. `2.8.5` (Nov 2009) seems to be the one that was shipped with original RHEL 6.0 release (Nov 2010) based on the RHEL6 linked document.

Yet there is [RHEL 6 documentation that actually shows output with GCM support](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/security_guide/sec-working_with_cipher_suites_in_gnutls) but there has been [no new releases since RHEL 6.10 (June 2018)](https://en.wikipedia.org/wiki/Red_Hat_Enterprise_Linux#RHEL_6).

---

On a slight tangent, [you later mention in another issue](https://github.com/mozilla/ssl-config-generator/issues/85#issuecomment-592939704) about OpenSSL supporting TLS 1.2 in RHEL 6. That [requires RHEL 6.5 (Nov 2013) minimum](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html-single/6.5_release_notes/index), or [6.8 (May 2016) for proper ECC support with other software such as Java 7 and 8](https://security.stackexchange.com/a/136515). I thought perhaps that meant similar improvement might have happened with GnuTLS, but supposedly that isn't the case.

No actual need for this myself, just looking into the varying cipher suite support and figured that anyone else that stumbles across this might like the additional details.
